src/learning/lda.rs

```diff
@@ -0,0 +1,250 @@
+    alpha: f64,
+    beta: f64,
+}
+
+/// An object which holds the results of Gibbs Sampling.
+///
+/// This object can then be used to get the distrbutions of
+/// topics over documents and words over topics
+#[derive(Debug)]
+pub struct LDAResult {
+    document_topic_count: Matrix<f64>,
+    topic_word_count: Matrix<f64>,
+    // The two vectors are simply used to reduce the amount of calculation that must be
+    // performed during the conditional probability step
+    topic_total_by_document: Vector<f64>,
+    word_total_by_topic: Vector<f64>,
+    alpha: f64,
+    beta: f64,
+}
+
+impl Default for LDA {
+    fn default() -> LDA {
+        LDA {
+            topic_count: 10,
+            alpha: 0.1,
+            beta: 0.1
+        }
+    }
+}
+
+impl LDA {
+    /// Creates a new object for finding LDA.
+    /// alpha and beta are the symmetric priors for the algorithm.
+    ///
+    /// If you don't know what to use, try alpha = 50/topic_count and beta = 0.01
+    pub fn new(topic_count: usize, alpha: f64, beta: f64) -> LDA {
+        LDA {
+            topic_count: topic_count,
+            alpha: alpha,
+            beta: beta
+        }
+    }
+}
+
+impl LDAResult {
+    fn new(input: &Matrix<usize>, topic_count: usize, alpha: f64, beta: f64) -> (LDAResult, Vec<Vec<usize> >) {
+        let document_count = input.rows();
+        let vocab_count = input.cols();
+        let mut topics = Vec::with_capacity(document_count);
+        let mut result = LDAResult {
+            document_topic_count: Matrix::zeros(document_count, topic_count),
+            topic_word_count: Matrix::zeros(topic_count, vocab_count),
+            topic_total_by_document: Vector::zeros(document_count),
+            word_total_by_topic: Vector::zeros(topic_count,),
+            alpha: alpha,
+            beta: beta
+        };
+
+        // For each word in each document, randomly assign it to a topic and update the counts
+        let mut rng = thread_rng();
+        for (document, row) in input.row_iter().enumerate() {
+            let mut document_topics = Vec::with_capacity(row.sum());
+            for (word, word_count) in row.iter().enumerate() {
+                for _ in 0..*word_count{
+                    let topic = rng.gen_range(0, topic_count);
+                    result.document_topic_count[[document, topic]] += 1.0;
+                    result.topic_total_by_document[document] += 1.0;
+                    result.topic_word_count[[topic, word]] += 1.0;
+                    result.word_total_by_topic[topic] += 1.0;
+                    document_topics.push(topic);
+                }
+            }
+            topics.push(document_topics);
+        }
+        (result, topics)
+    }
+
+    /// Find the distribution of words over topics.  This gives a matrix where the rows are
+    /// topics and the columns are words.  Each entry (topic, word) gives the probability of
+    /// word given topic.
+    pub fn phi(&self) -> Matrix<f64> {
```

Comment:
This name seems somewhat dependent on related literature. I worry that unfamiliar users (like myself) would not know where to look for this function. Maybe it should be renamed `distr_of_words`?